// lance/utils.js
// Generated by CoffeeScript 1.9.3
(function() {
  var Promise, clone, coroutiner, merge, ref, typeOf, utils;

  Promise = require('bluebird');

  ref = utils = require('lutils'), merge = ref.merge, clone = ref.clone, typeOf = ref.typeOf;

  coroutiner = new require('coroutiner').Coroutiner({
    prototype: true
  });

  coroutiner.transformer = Promise.coroutine;


  /*
  	Clone lutils then extend the new object
   */

  module.exports = utils = merge({
    format: require('./utils/format'),
    prettyError: require('./utils/prettyError'),
    exploreDir: require('./utils/exploreDir'),
    coroutiner: coroutiner
  }, utils);

  utils.inspect = function(vari, depth, showHidden, colors) {
    if (depth == null) {
      depth = 6;
    }
    if (showHidden == null) {
      showHidden = true;
    }
    if (colors == null) {
      colors = true;
    }
    return require('util').inspect(vari, {
      depth: depth,
      showHidden: showHidden,
      colors: colors
    });
  };

}).call(this);
